File: samples/post-fx/PostFX.cpp

////////////////////////////////////////////////////////////
// Headers
////////////////////////////////////////////////////////////
#include <SFML/Graphics.hpp>
#include <map>
#include <math.h>

void DisplayError();


////////////////////////////////////////////////////////////
/// Entry point of application
///
/// \return Application exit code
///
////////////////////////////////////////////////////////////
int main()
{
    // Check that the system can use post effects
    if (sf::PostFX::CanUsePostFX() == false)
    {
        DisplayError();
        return EXIT_SUCCESS;
    }

    // Create the main window
    sf::RenderWindow window(sf::VideoMode(800, 600), "SFML PostFX");

    // Load a background image to display
    sf::Image backgroundImage;
    if (!backgroundImage.LoadFromFile("datas/post-fx/background.jpg"))
        return EXIT_FAILURE;
    sf::Sprite background(backgroundImage);

    // Load a sprite which we'll move into the scene
    sf::Image entityImage;
    if (!entityImage.LoadFromFile("datas/post-fx/sprite.png"))
        return EXIT_FAILURE;
    sf::Sprite entity(entityImage);

    // Load the text font
    sf::Font font;
    if (!font.LoadFromFile("datas/post-fx/cheeseburger.ttf"))
        return EXIT_FAILURE;

    // Load the image needed for the wave effect
    sf::Image waveImage;
    if (!waveImage.LoadFromFile("datas/post-fx/wave.jpg"))
        return EXIT_FAILURE;

    // Load all effects
    std::map<std::string, sf::PostFX> effects;
    if (!effects["nothing"].LoadFromFile("datas/post-fx/nothing.sfx"))   return EXIT_FAILURE;
    if (!effects["blur"].LoadFromFile("datas/post-fx/blur.sfx"))         return EXIT_FAILURE;
    if (!effects["colorize"].LoadFromFile("datas/post-fx/colorize.sfx")) return EXIT_FAILURE;
    if (!effects["fisheye"].LoadFromFile("datas/post-fx/fisheye.sfx"))   return EXIT_FAILURE;
    if (!effects["wave"].LoadFromFile("datas/post-fx/wave.sfx"))         return EXIT_FAILURE;
    if (!effects["pixelate"].LoadFromFile("datas/post-fx/pixelate.sfx")) return EXIT_FAILURE;
    std::map<std::string, sf::PostFX>::iterator currentEffect = effects.find("nothing");

    // Do specific initializations
    effects["nothing"].SetTexture("framebuffer", NULL);
    effects["blur"].SetTexture("framebuffer", NULL);
    effects["blur"].SetParameter("offset", 0.f);
    effects["colorize"].SetTexture("framebuffer", NULL);
    effects["colorize"].SetParameter("color", 1.f, 1.f, 1.f);
    effects["fisheye"].SetTexture("framebuffer", NULL);
    effects["wave"].SetTexture("framebuffer", NULL);
    effects["wave"].SetTexture("wave", &waveImage);
    effects["pixelate"].SetTexture("framebuffer", NULL);

    // Define a string for displaying current effect description
    sf::String curFXStr;
    curFXStr.SetText("Current effect is \"" + currentEffect->first + "\"");
    curFXStr.SetFont(font);
    curFXStr.SetPosition(20.f, 0.f);
    curFXStr.SetColor(sf::Color(150, 70, 110));

    // Define a string for displaying help
    sf::String infoStr;
    infoStr.SetText("Move your mouse to change the effect parameters\nPress numpad + and - to change effect\nWarning : some effects may not work\ndepending on your graphics card");
    infoStr.SetFont(font);
    infoStr.SetPosition(20.f, 460.f);
    infoStr.SetColor(sf::Color(200, 100, 150));

    // Create a clock to measure the total time elapsed
    sf::Clock clock;

    // Start the game loop
    while (window.IsOpened())
    {
        // Process events
        sf::Event event;
        while (window.GetEvent(event))
        {
            // Close window : exit
            if (event.Type == sf::Event::Closed)
                window.Close();

            if (event.Type == sf::Event::KeyPressed)
            {
                // Escape key : exit
                if (event.Key.Code == sf::Key::Escape)
                    window.Close();

                // Add key : next effect
                if (event.Key.Code == sf::Key::Add)
                {
                    currentEffect++;
                    if (currentEffect == effects.end())
                        currentEffect = effects.begin();
                    curFXStr.SetText("Current effect is \"" + currentEffect->first + "\"");
                }

                // Subtract key : previous effect
                if (event.Key.Code == sf::Key::Subtract)
                {
                    if (currentEffect == effects.begin())
                        currentEffect = effects.end();
                    currentEffect--;
                    curFXStr.SetText("Current effect is \"" + currentEffect->first + "\"");
                }
            }
        }

        // Get the mouse position in the range [0, 1]
        float mouseX = window.GetInput().GetMouseX() / static_cast<float>(window.GetWidth());
        float mouseY = window.GetInput().GetMouseY() / static_cast<float>(window.GetHeight());

        // Update the current effect
        if      (currentEffect->first == "blur")     currentEffect->second.SetParameter("offset", mouseX * mouseY * 0.05f);
        else if (currentEffect->first == "colorize") currentEffect->second.SetParameter("color", 0.3f, mouseX, mouseY);
        else if (currentEffect->first == "fisheye")  currentEffect->second.SetParameter("mouse", mouseX, 1.f - mouseY);
        else if (currentEffect->first == "wave")     currentEffect->second.SetParameter("offset", mouseX, mouseY);
        else if (currentEffect->first == "pixelate") currentEffect->second.SetParameter("mouse", mouseX, mouseY);

        // Animate the sprite
        float entityX = (cos(clock.GetElapsedTime() * 1.3f) + 1.2f) * 300;
        float entityY = (cos(clock.GetElapsedTime() * 0.8f) + 1.2f) * 200;
        entity.SetPosition(entityX, entityY);
        entity.Rotate(window.GetFrameTime() * 100);

        // Clear the window
        window.Clear();

        // Draw background, sprite and apply the post-fx
        window.Draw(background);
        window.Draw(entity);
        window.Draw(currentEffect->second);

        // Draw interface strings
        window.Draw(curFXStr);
        window.Draw(infoStr);

        // Finally, display the rendered frame on screen
        window.Display();
    }

    return EXIT_SUCCESS;
}


////////////////////////////////////////////////////////////
/// Fonction called when the post-effects are not supported ;
/// Display an error message and wait until the user exits
///
////////////////////////////////////////////////////////////
void DisplayError()
{
    // Create the main window
    sf::RenderWindow window(sf::VideoMode(800, 600), "SFML PostFX");

    // Define a string for displaying the error message
    sf::String error("Sorry, your system doesn't support post-effects");
    error.SetPosition(100.f, 250.f);
    error.SetColor(sf::Color(200, 100, 150));

    // Start the game loop
    while (window.IsOpened())
    {
        // Process events
        sf::Event event;
        while (window.GetEvent(event))
        {
            // Close window : exit
            if (event.Type == sf::Event::Closed)
                window.Close();

            // Escape key : exit
            if ((event.Type == sf::Event::KeyPressed) && (event.Key.Code == sf::Key::Escape))
                window.Close();
        }

        // Clear the window
        window.Clear();

        // Draw the error message
        window.Draw(error);

        // Finally, display the rendered frame on screen
        window.Display();
    }
}

File: dotnet/samples/post-fx/PostFx.cs
using System;
using System.Collections.Generic;
using SFML;
using SFML.Graphics;
using SFML.Window;

namespace sample_postfx
{
    static class Program
    {
        private static Dictionary<string, PostFx>            effects;
        private static Dictionary<string, PostFx>.Enumerator currentEffect;
        private static String2D                              curFXStr;

        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        static void Main()
        {
            // Create the main window
            RenderWindow window = new RenderWindow(new VideoMode(800, 600), "SFML.Net PostFX");

            // Setup event handlers
            window.Closed     += new EventHandler(OnClosed);
            window.KeyPressed += new EventHandler<KeyEventArgs>(OnKeyPressed);

            // Check that the system can use post effects
            if (PostFx.CanUsePostFX == false)
            {
                DisplayError(window);
                return;
            }

            // Load a background image to display
            Sprite background = new Sprite(new Image("datas/post-fx/background.jpg"));

            // Load a sprite which we'll move into the scene
            Sprite entity = new Sprite(new Image("datas/post-fx/sprite.png"));

            // Load the text font
            Font cheeseburger = new Font("datas/post-fx/cheeseburger.ttf");

            // Load the image needed for the wave effect
            Image waveImage = new Image("datas/post-fx/wave.jpg");

            // Load all effects
            effects = new Dictionary<string, PostFx>();
            effects["nothing"]  = new PostFx("datas/post-fx/nothing.sfx");
            effects["blur"]     = new PostFx("datas/post-fx/blur.sfx");
            effects["colorize"] = new PostFx("datas/post-fx/colorize.sfx");
            effects["fisheye"]  = new PostFx("datas/post-fx/fisheye.sfx");
            effects["wave"]     = new PostFx("datas/post-fx/wave.sfx");
            effects["pixelate"] = new PostFx("datas/post-fx/pixelate.sfx");
            currentEffect = effects.GetEnumerator();
            currentEffect.MoveNext();

            // Do specific initializations
            effects["nothing"].SetTexture("framebuffer", null);
            effects["blur"].SetTexture("framebuffer", null);
            effects["blur"].SetParameter("offset", 0.0F);
            effects["colorize"].SetTexture("framebuffer", null);
            effects["colorize"].SetParameter("color", 1.0F, 1.0F, 1.0F);
            effects["fisheye"].SetTexture("framebuffer", null);
            effects["wave"].SetTexture("framebuffer", null);
            effects["wave"].SetTexture("wave", waveImage);
            effects["pixelate"].SetTexture("framebuffer", null);

            // Define a string for displaying current effect description
            curFXStr = new String2D();
            curFXStr.Text = "Current effect is \"" + currentEffect.Current.Key + "\"";
            curFXStr.Font = cheeseburger;
            curFXStr.Position = new Vector2(20.0F, 0.0F);
            curFXStr.Color = new Color(150, 70, 110);

            // Define a string for displaying help
            String2D infoStr = new String2D();
            infoStr.Text = "Move your mouse to change the effect parameters\nPress numpad + to change effect\nWarning : some effects may not work\ndepending on your graphics card";
            infoStr.Font = cheeseburger;
            infoStr.Position = new Vector2(20.0F, 460.0F);
            infoStr.Color = new Color(200, 100, 150);

            // Start the game loop
            float time = 0.0F;
            while (window.IsOpened())
            {
                // Process events
                window.DispatchEvents();

                // Get the mouse position in the range [0, 1]
                float x = window.Input.GetMouseX() / (float)window.Width;
                float y = window.Input.GetMouseY() / (float)window.Height;

                // Update the current effect
                if      (currentEffect.Current.Key == "blur")     currentEffect.Current.Value.SetParameter("offset", x * y * 0.1f);
                else if (currentEffect.Current.Key == "colorize") currentEffect.Current.Value.SetParameter("color", 0.3f, x, y);
                else if (currentEffect.Current.Key == "fisheye")  currentEffect.Current.Value.SetParameter("mouse", x, 1.0F - y);
                else if (currentEffect.Current.Key == "wave")     currentEffect.Current.Value.SetParameter("offset", x, y);
                else if (currentEffect.Current.Key == "pixelate") currentEffect.Current.Value.SetParameter("mouse", x, y);

                // Animate the sprite
                time += window.GetFrameTime();
                float entityX = (float)(Math.Cos(time * 1.3) + 1.2) * 300;
                float entityY = (float)(Math.Cos(time * 0.8) + 1.2) * 200;
                entity.Position = new Vector2(entityX, entityY);
                entity.Rotation = time * 100;

                // Clear the window
                window.Clear();

                // Draw background, the sprite and apply the post-fx
                window.Draw(background);
                window.Draw(entity);
                window.Draw(currentEffect.Current.Value);

                // Draw interface strings
                window.Draw(curFXStr);
                window.Draw(infoStr);

                // Finally, display the rendered frame on screen
                window.Display();
            }
        }

        /// <summary>
        /// Fonction called when the post-effects are not supported ;
        /// Display an error message and wait until the user exits
        /// </summary>
        private static void DisplayError(RenderWindow window)
        {
            // Define a string for displaying the error message
            String2D errorStr = new String2D("Sorry, your system doesn't support post-effects");
            errorStr.Position = new Vector2(100.0F, 250.0F);
            errorStr.Color = new Color(200, 100, 150);

            // Start the game loop
            while (window.IsOpened())
            {
                // Process events
                window.DispatchEvents();

                // Clear the window
                window.Clear();

                // Draw the error message
                window.Draw(errorStr);

                // Finally, display the rendered frame on screen
                window.Display();
            }
        }

        /// <summary>
        /// Function called when the window is closed
        /// </summary>
        static void OnClosed(object sender, EventArgs e)
        {
            RenderWindow window = (RenderWindow)sender;
            window.Close();
        }

        /// <summary>
        /// Function called when a key is pressed
        /// </summary>
        static void OnKeyPressed(object sender, KeyEventArgs e)
        {
            RenderWindow window = (RenderWindow)sender;
            if (e.Code == KeyCode.Escape)
            {
                // Close the window
                window.Close();
            }
            else if (e.Code == KeyCode.Add)
            {
                // Advance to the next effect
                if (currentEffect.MoveNext() == false)
                {
                    currentEffect = effects.GetEnumerator();
                    currentEffect.MoveNext();
                }
                curFXStr.Text = "Current effect is \"" + currentEffect.Current.Key + "\"";
            }
        }
    }
}
